bee-protocol/src/workers/peer/packet_handler.rs

```diff
@@ -0,0 +1,333 @@
+            // The handler should read a header first.
+            state: ReadState::Header,
+            address,
+        }
+    }
+    /// Fetch the header and payload of a packet.
+    ///
+    /// This method only returns `None` if a shutdown signal is received.
+    pub(super) async fn fetch_packet(&mut self) -> Option<(HeaderPacket, &[u8])> {
+        // loop until we can return the header and payload
+        loop {
+            match &self.state {
+                // Read a header.
+                ReadState::Header => {
+                    // We need `HEADER_SIZE` bytes to read a header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, HEADER_SIZE)
+                        .await?;
+                    trace!("[{}] Reading Header...", self.address);
+                    let header = HeaderPacket::from_bytes(bytes);
+                    // Now we are ready to read a payload.
+                    self.state = ReadState::Payload(header);
+                }
+                // Read a payload.
+                ReadState::Payload(header) => {
+                    // We read the quantity of bytes stated by the header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, header.packet_length.into())
+                        .await?;
+                    // FIXME: Avoid this clone
+                    let header = header.clone();
+                    // Now we are ready to read the next packet's header.
+                    self.state = ReadState::Header;
+                    // We return the current packet's header and payload.
+                    return Some((header, bytes));
+                }
+            }
+        }
+    }
+}
+
+// An event handler.
+//
+// This type takes care of actually receiving the events and appending them to an inner buffer so
+// they can be used seamlessly by the `PacketHandler`.
+struct EventHandler {
+    receiver: EventRecv,
+    buffer: Vec<u8>,
+    offset: usize,
+}
+
+impl EventHandler {
+    /// Create a new event handler from an event receiver.
+    fn new(receiver: EventRecv) -> Self {
+        Self {
+            receiver,
+            buffer: vec![],
+            offset: 0,
+        }
+    }
+
+    /// Push a new event into the buffer.
+    ///
+    /// This method also removes the `..self.offset` range from the buffer and sets the offset back
+    /// to zero. Which means that this should only be called when the buffer is empty or when there
+    /// are not enough bytes to read a new header or payload.
+    fn push_event(&mut self, mut bytes: Vec<u8>) {
+        // Remove the already read bytes from the buffer.
+        self.buffer = self.buffer.split_off(self.offset);
+        // Reset the offset.
+        self.offset = 0;
+        // Append the bytes of the new event
+        if self.buffer.is_empty() {
```

Comment:
why not `self.buffer.append(&mut bytes)` instead of walking around the bush?

Reply:
What do you mean ?

Reply:
we are replacing the buffer if the current buffer is empty and extending it if not. Why not just extend it?

Reply:
I think you wrote this code, wouldn't it be because the `=` would do a move of the vector while the extend would do an allocation ?

Reply:
Christian with the longer telomeres is a dummy :trollface:

Just `self.buffer.append(&mut bytes)` optimizes better apparently.